presto-accumulo/src/main/java/com/facebook/presto/accumulo/index/ColumnCardinalityCache.java

```diff
@@ -99,250 +117,295 @@ public synchronized void deleteCache(String schema, String table)
         }
 
-        @SuppressWarnings("unchecked")
         @Override
-        public Map<Range, Long> loadAll(Iterable<? extends Range> keys)
+        public Map<CacheKey, Long> loadAll(@Nonnull Iterable<? extends CacheKey> keys)
                 throws Exception
         {
-            LOG.debug("Loading %s exact ranges from Accumulo", ((Collection<Range>) keys).size());
-
-            // Create batch scanner for querying all ranges
-            BatchScanner scanner = connector.createBatchScanner(metricsTable, auths, 10);
-            scanner.setRanges((Collection<Range>) keys);
-            scanner.fetchColumn(columnFamily, Indexer.CARDINALITY_CQ_AS_TEXT);
-
-            // Create a new map to hold our cardinalities for each range, returning a default of zero for each non-existent Key
-            Map<Range, Long> rangeValues = new MapDefaultZero();
-            for (Entry<Key, Value> entry : scanner) {
-                rangeValues.put(
-                        Range.exact(entry.getKey().getRow()),
-                        Long.parseLong(entry.getValue().toString()));
+            int size = Iterables.size(keys);
+            if (size == 0) {
+                return ImmutableMap.of();
             }
 
-            scanner.close();
-            return rangeValues;
-        }
+            LOG.debug("Loading %s exact ranges from Accumulo", size);
 
-        /**
-         * We extend HashMap here and override get to return a value of zero if the key is not in the map.
-         * This mitigates the CacheLoader InvalidCacheLoadException if loadAll fails to return a value for a given key,
-         * which occurs when there is no key in Accumulo.
-         */
-        public class MapDefaultZero
-                extends HashMap<Range, Long>
-        {
-            @Override
-            public Long get(Object key)
-            {
-                // Get the key from our map overlord
-                Long value = super.get(key);
-
-                // Return zero if null
-                return value == null ? 0 : value;
+            // In order to simplify the implementation, we are making a (safe) assumption
+            // that the CacheKeys will all contain the same combination of schema/table/family/qualifier
+            // This is asserted with the below implementation error just to make sure
+            CacheKey anyKey = stream(keys).findAny().get();
+            if (stream(keys).anyMatch(k -> !k.getSchema().equals(anyKey.getSchema()) || !k.getTable().equals(anyKey.getTable()) || !k.getFamily().equals(anyKey.getFamily()) || !k.getQualifier().equals(anyKey.getQualifier()))) {
+                throw new PrestoException(FUNCTION_IMPLEMENTATION_ERROR, "loadAll called with a non-homogeneous collection of cache keys");
+            }
+
+            Map<Range, CacheKey> rangeToKey = stream(keys).collect(Collectors.toMap(CacheKey::getRange, Function.identity()));
+            LOG.debug("rangeToKey size is " + rangeToKey.size());
```

Comment:
This should pass a format string to avoid constructing the log message when the log level is not enabled.